contracts/distributionroles.js

```diff
@@ -0,0 +1,407 @@
+/* eslint no-underscore-dangle: ["error", { "allow": ["_id"] }] */
+/* global actions, api */
+
+const DUST_PCT = 0.01;
+const MAX_RECIPIENTS = 40;
+const VOTES_PER_ROLE = 1;
+
+actions.createSSC = async () => {
+  const tableExists = await api.db.tableExists('batches');
+  if (tableExists === false) {
+    await api.db.createTable('batches');
+    await api.db.createTable('params');
+
+    const params = {};
+    params.distCreationFee = '750';
+    params.distUpdateFee = '500';
+    await api.db.insert('params', params);
+  }
+};
+
+actions.updateParams = async (payload) => {
+  if (api.sender !== api.owner) return;
+
+  const {
+    distCreationFee,
+    distUpdateFee,
+  } = payload;
+
+  const params = await api.db.findOne('params', {});
+
+  if (distCreationFee) {
+    if (!api.assert(typeof distCreationFee === 'string' && !api.BigNumber(distCreationFee).isNaN() && api.BigNumber(distCreationFee).gte(0), 'invalid distCreationFee')) return;
+    params.distCreationFee = distCreationFee;
+  }
+  if (distUpdateFee) {
+    if (!api.assert(typeof distUpdateFee === 'string' && !api.BigNumber(distUpdateFee).isNaN() && api.BigNumber(distUpdateFee).gte(0), 'invalid distUpdateFee')) return;
+    params.distUpdateFee = distUpdateFee;
+  }
+
+  await api.db.update('params', params);
+};
+
+async function validateRoles(roles) {
+  if (!api.assert(roles && Array.isArray(roles), 'roles must be an array')) return false;
+  if (!api.assert(roles.length >= 1, 'specify at least one role')) return false;
+
+  const rolesNames = new Set();
+  let roleTotalShare = 0;
+  let roleTotalPrimary = 0;
+  for (let i = 0; i < roles.length; i += 1) {
+    const rolesConfig = roles[i];
+    if (!api.assert(rolesConfig && rolesConfig.name
+      && typeof (rolesConfig.name) === 'string' && rolesConfig.name.length <= 255, 'roles name invalid')) return false;
+
+    if (!api.assert(!rolesNames.has(rolesConfig.name), 'roles cannot have duplicate names')) return false;
+    rolesNames.add(rolesConfig.name);
+
+    if (!api.assert(rolesConfig && rolesConfig.description
+      && typeof (rolesConfig.description) === 'string' && rolesConfig.description.length <= 255, 'roles description invalid')) return false;
+
+    if (!api.assert(Number.isInteger(rolesConfig.pct)
+      && rolesConfig.pct >= 1 && rolesConfig.pct <= 100, 'roles pct must be an integer from 1 to 100')) return false;
+
+    if (!api.assert(Number.isInteger(rolesConfig.pct)
+      && rolesConfig.primary >= 1 && rolesConfig.primary <= MAX_RECIPIENTS, `roles primary must be an integer from 1 to ${MAX_RECIPIENTS}`)) return false;
+
+    roleTotalShare += rolesConfig.pct;
+    roleTotalPrimary += rolesConfig.primary;
+  }
+  if (!api.assert(roleTotalShare === 100, 'roles pct must total 100')) return false;
+  if (!api.assert(roleTotalPrimary <= MAX_RECIPIENTS, `total of roles primary must not exceed ${MAX_RECIPIENTS}`)) return false;
+  return true;
+}
+
+async function updateStakeWeight(distOrId) {
```

Comment:
Other comment is that this will have overlap with the witness contract. Have we considered repurposing some of that? (E.g. backup rotation mechanism instead of proportional backup distribution)

Reply:
Some of the code was repurposed from the witness contract. I actually started with an incremental approach, I need to look at it again, but for some reason I was working with the premise not to introduce additional ticks in block or edit other contracts (tokens) @bt-cryptomancer. With that, it could mimic the mining contract.

As for the rotation mechanism, I don't believe that's been discussed. Proportional seems to me easier to understand for the orgs that may use this. The example given was for paying contributors to a project, rather than time-sensitive witnessing of blocks.

Reply:
Yeah I'm pretty worried about how intensive this computation is as the number of stakers increase. I would definitely prefer modifying tokens contract. The other way we can do it is to generalize the mining contract to create a hidden pool that only accounts for delegation (delegated mining power = voting) and read out the mining power of this hidden pool. Perhaps we should brainstorm this out. The reason I suggest this is because we already do this kind of incremental computation in the mining contract